sources/api/netdog/src/net_config/devices/mod.rs

```diff
@@ -0,0 +1,128 @@
+//! The devices module contains all the types of network devices that `netdog` supports. These are
+//! intended to be the structs used for net.toml deserialization including the validation logic for
+//! each device.
+
+pub(crate) mod bonding;
+pub(crate) mod interface;
+pub(crate) mod vlan;
+
+use super::{error, Result, Validate};
+use crate::net_config::{Dhcp4ConfigV1, Dhcp6ConfigV1, RouteV1, StaticConfigV1};
+use bonding::NetBondV1;
+use interface::NetInterfaceV2;
+use serde::Deserialize;
+use vlan::NetVlanV1;
+
+#[derive(Debug, Deserialize)]
+#[serde(untagged)]
+pub(crate) enum NetworkDeviceV1 {
+    Interface(NetInterfaceV2),
+    BondDevice(NetBondV1),
+    VlanDevice(NetVlanV1),
+}
+
+impl NetworkDeviceV1 {
+    pub(crate) fn primary(&self) -> Option<bool> {
+        match self {
+            Self::Interface(i) => i.primary,
+            Self::BondDevice(i) => i.primary,
+            Self::VlanDevice(i) => i.primary,
+        }
+    }
+}
+
+impl Validate for NetworkDeviceV1 {
+    fn validate(&self) -> Result<()> {
+        match self {
+            Self::Interface(config) => config.validate()?,
+            Self::BondDevice(config) => config.validate()?,
+            Self::VlanDevice(config) => config.validate()?,
+        }
+        Ok(())
+    }
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(untagged)]
+pub(crate) enum DeviceType {
+    #[serde(rename = "bond")]
+    Bond,
+    #[serde(rename = "vlan")]
+    Vlan,
+}
+
+pub(crate) trait HasIpAddressing {
+    fn has_static(&self) -> bool;
+
+    fn validate_static4(&self) -> Result<()>;
+    fn validate_static6(&self) -> Result<()>;
+
+    fn has_dhcp(&self) -> bool;
+    fn has_routes(&self) -> bool;
+}
+
+pub(crate) fn validate_addressing<D>(device: D) -> Result<()>
+where
+    D: HasIpAddressing,
+{
+    if !device.has_dhcp() && !device.has_static() {
+        return error::InvalidNetConfigSnafu {
+            reason: "each interface must configure dhcp and/or static addresses",
+        }
+        .fail();
+    }
+
+    // wicked doesn't support static routes with dhcp
+    if device.has_dhcp() && device.has_routes() {
+        return error::InvalidNetConfigSnafu {
+            reason: "static routes are not supported with dhcp",
+        }
+        .fail();
+    }
+
+    if device.has_routes() && !device.has_static() {
+        return error::InvalidNetConfigSnafu {
+            reason: "interfaces must set static addresses in order to use routes",
+        }
+        .fail();
+    }
+
+    // call into struct for access to fields for validation
+    device.validate_static4()?;
+    device.validate_static6()?;
+
+    Ok(())
+}
+
+// For all devices that have IP Addressing available, generate the trait implementation
+macro_rules! generate_addressing_validation {
+    ($name:ty) => {
```

Comment:
The macro and trait look OK to me. However, the generated method could potentially also be invoked in the custom deserializer, which when combined with everything else could eliminate the need for the `Validate` trait and invocation.

Reply:
Yeah - that's probably a good move!  Eliminates some boilerplate as well.

Reply:
This makes sense to me. We can do the clean up on this macro call with the rest of the move to Deserialize and that should remove the need for validate all in one change.

Reply:
https://github.com/bottlerocket-os/bottlerocket/issues/2657 covers this work.